// @codemod/matchers/build/matchers/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.tupleOf = exports.matcher = exports.or = exports.oneOf = exports.Matcher = exports.function = exports.Function = exports.fromCapture = exports.containerOf = exports.CapturedMatcher = exports.capture = exports.arrayOf = exports.anything = exports.anyString = exports.anyStatement = exports.anyNumber = exports.anyNode = exports.anyList = exports.anyExpression = void 0;
var anyExpression_1 = require("./anyExpression");
Object.defineProperty(exports, "anyExpression", { enumerable: true, get: function () { return anyExpression_1.anyExpression; } });
var anyList_1 = require("./anyList");
Object.defineProperty(exports, "anyList", { enumerable: true, get: function () { return anyList_1.anyList; } });
var anyNode_1 = require("./anyNode");
Object.defineProperty(exports, "anyNode", { enumerable: true, get: function () { return anyNode_1.anyNode; } });
var anyNumber_1 = require("./anyNumber");
Object.defineProperty(exports, "anyNumber", { enumerable: true, get: function () { return anyNumber_1.anyNumber; } });
var anyStatement_1 = require("./anyStatement");
Object.defineProperty(exports, "anyStatement", { enumerable: true, get: function () { return anyStatement_1.anyStatement; } });
var anyString_1 = require("./anyString");
Object.defineProperty(exports, "anyString", { enumerable: true, get: function () { return anyString_1.anyString; } });
var anything_1 = require("./anything");
Object.defineProperty(exports, "anything", { enumerable: true, get: function () { return anything_1.anything; } });
var arrayOf_1 = require("./arrayOf");
Object.defineProperty(exports, "arrayOf", { enumerable: true, get: function () { return arrayOf_1.arrayOf; } });
var capture_1 = require("./capture");
Object.defineProperty(exports, "capture", { enumerable: true, get: function () { return capture_1.capture; } });
Object.defineProperty(exports, "CapturedMatcher", { enumerable: true, get: function () { return capture_1.CapturedMatcher; } });
var containerOf_1 = require("./containerOf");
Object.defineProperty(exports, "containerOf", { enumerable: true, get: function () { return containerOf_1.containerOf; } });
var fromCapture_1 = require("./fromCapture");
Object.defineProperty(exports, "fromCapture", { enumerable: true, get: function () { return fromCapture_1.fromCapture; } });
var function_1 = require("./function");
Object.defineProperty(exports, "Function", { enumerable: true, get: function () { return function_1.Function; } });
Object.defineProperty(exports, "function", { enumerable: true, get: function () { return function_1.function; } });
__exportStar(require("./generated"), exports);
var Matcher_1 = require("./Matcher");
Object.defineProperty(exports, "Matcher", { enumerable: true, get: function () { return Matcher_1.Matcher; } });
var oneOf_1 = require("./oneOf");
Object.defineProperty(exports, "oneOf", { enumerable: true, get: function () { return oneOf_1.oneOf; } });
var or_1 = require("./or");
Object.defineProperty(exports, "or", { enumerable: true, get: function () { return or_1.or; } });
var predicate_1 = require("./predicate");
Object.defineProperty(exports, "matcher", { enumerable: true, get: function () { return predicate_1.predicate; } });
__exportStar(require("./slice"), exports);
var tupleOf_1 = require("./tupleOf");
Object.defineProperty(exports, "tupleOf", { enumerable: true, get: function () { return tupleOf_1.tupleOf; } });
